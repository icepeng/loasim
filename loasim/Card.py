from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
