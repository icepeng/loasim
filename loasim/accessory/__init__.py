from loasim.accessory.builtin import lostark_accessory_repository
from loasim.accessory.base import AccessoryState
