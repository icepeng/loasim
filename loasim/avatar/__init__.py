from loasim.avatar.base import AvatarState
from loasim.avatar.builtin import lostark_avatar_repository
