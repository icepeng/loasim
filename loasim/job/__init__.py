from loasim.job.artist import lostark_job_artist
from loasim.job.base import Job, SkillSpecification, SkillState, Tripod
from loasim.job.blade import lostark_job_blade
from loasim.job.hawkeye import lostark_job_hawkeye
from loasim.job.scouter import lostark_job_scouter

job_dict = {
    "artist": lostark_job_artist,
    "blade": lostark_job_blade,
    "hawkeye": lostark_job_hawkeye,
    "scouter": lostark_job_scouter,
}


def get_job(job_name: str) -> Job:
    job = job_dict[job_name]
    if job is None:
        raise ValueError(f"{job_name} is not supported")
    return job
