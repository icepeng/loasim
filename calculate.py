from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
