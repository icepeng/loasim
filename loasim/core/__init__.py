from loasim.core.buff import (
    BuffManager,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import Skill, SkillSpecification, Tripod
from loasim.core.stat import InternalStat, Stat
