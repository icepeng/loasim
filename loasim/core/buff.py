from __future__ import annotations

from typing import Callable, Dict, List, Union

from pydantic import BaseModel

from loasim.core.skill import Skill
from loasim.core.stat import InternalStat, Stat


class BuffState(BaseModel):
    onoff: bool
    stack: int = 0


class AbstractBuff(BaseModel):
    name: str

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ) -> Stat:
        raise NotImplementedError()


class OnoffBuff(AbstractBuff):
    name: str
    stat: Stat

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        buff_state = state.get(self.name)
        if buff_state is None:
            raise ValueError(f"{self.name} is not found in state")
        if buff_state.onoff:
            return self.stat
        return Stat()


class StaticBuff(AbstractBuff):
    name: str
    stat: Stat

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        return self.stat


class StackBuff(AbstractBuff):
    name: str
    stat_fn: Union[
        Callable[[int], Stat], Callable[[int], Stat]
    ]  # mypy #5485 workaround

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        buff_state = state.get(self.name)
        if buff_state is None:
            raise ValueError(f"{self.name} is not found in state")
        if buff_state.onoff:
            return self.stat_fn(buff_state.stack)
        return Stat()


class SkillBuff(AbstractBuff):
    name: str
    stat_fn: Union[
        Callable[[Skill], Stat], Callable[[Skill], Stat]
    ]  # mypy #5485 workaround

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        return self.stat_fn(skill)


class StatBuff(AbstractBuff):
    name: str
    stat_fn: Union[
        Callable[[Stat], Stat], Callable[[Stat], Stat]
    ]  # mypy #5485 workaround

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        return self.stat_fn(stat)


class InternalStatOnoffBuff(AbstractBuff):
    name: str
    stat_fn: Union[
        Callable[[InternalStat], Stat], Callable[[InternalStat], Stat]
    ]  # mypy #5485 workaround

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        stat: Stat,
        internal_stat: InternalStat,
    ):
        buff_state = state.get(self.name)
        if buff_state is None:
            raise ValueError(f"{self.name} is not found in state")
        if buff_state.onoff:
            return self.stat_fn(internal_stat)
        return Stat()


class BuffManager:
    def __init__(self, buff_list: List[AbstractBuff]) -> None:
        self._buffs: Dict[str, AbstractBuff] = {}
        self._stat_buffs: Dict[str, StatBuff] = {}
        for buff in buff_list:
            if isinstance(buff, StatBuff):
                self._stat_buffs[buff.name] = buff
            else:
                self._buffs[buff.name] = buff

    def get_stat(
        self,
        state: Dict[str, BuffState],
        skill: Skill,
        basis_stat: Stat,
        internal_stat: InternalStat,
    ) -> Stat:
        stat = Stat()
        for buff in self._buffs.values():
            stat = stat + buff.get_stat(state, skill, stat, internal_stat)

        for buff in self._stat_buffs.values():
            stat = stat + buff.get_stat(state, skill, stat + basis_stat, internal_stat)

        return stat
