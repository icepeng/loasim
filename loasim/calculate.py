from loguru import logger

from collections import defaultdict
from typing import Any, Callable, List, Dict, Literal, Tuple

from pydantic import BaseModel

from loasim.core import Skill, Enemy, BuffManager, InternalStat
from loasim.core.buff import Buff
from loasim.core.stat import Stat
from loasim.setitem import lostark_setitem_repository
from loasim.card import lostark_default_card_repository
from loasim.engraving import lostark_engraving_repository


class DealCycle(BaseModel):
    buff_list: List[Buff]
    skill_list: List[Tuple[Skill, Dict[str, Any], Literal["back", "head"] | None]]
    cycle_time: float


def calculate(
    internal_stat: InternalStat,
    weapon_pdamage: float,
    setitem_state: List[Tuple[str, int, int]],
    engraving_state: List[Tuple[str, int]],
    card_state: List[str],
    generator: Callable[[InternalStat], DealCycle],
    enemy: Enemy,
):
    deal_cycle = generator(internal_stat)
    base_stat = internal_stat.get_stat()
    weapon_stat = Stat(pdamage=weapon_pdamage)  # 무기 품질
    basis_stat = (
        base_stat
        + weapon_stat
        + lostark_setitem_repository.get_stat(setitem_state)
        + lostark_default_card_repository.get_stat(card_state)
    )

    logger.info(base_stat)
    logger.info(basis_stat)

    engraving_buffs = lostark_engraving_repository.get_buffs(engraving_state)
    buff_manager = BuffManager(engraving_buffs + deal_cycle.buff_list)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill, buff_state, backhead in deal_cycle.skill_list:
        buff_stat = buff_manager.get_stat(buff_state, skill)
        damage = skill.get_damage(enemy, backhead, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(deal_cycle.cycle_time)
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")
