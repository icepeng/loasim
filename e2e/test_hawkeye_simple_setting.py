from loasim.calculate import DealCycle, calculate
from loasim.core import Enemy, InternalStat
from loasim.core.buff import BuffState
from loasim.job import SkillState

state = {
    "피해 증폭": BuffState(onoff=True),
}

deal_cycle: DealCycle = [
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("샤프 슈터", state, None),
    ("스나이프", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("아토믹 애로우", state, None),
    ("샤프 슈터", state, None),
    ("스나이프", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("샤프 슈터", state, None),
    ("아토믹 애로우", state, None),
    ("스나이프", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("샤프 슈터", state, None),
    ("스나이프", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 해일", state, None),
    ("스나이프", state, None),
    ("샤프 슈터", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("스나이프", state, None),
    ("샤프 슈터", state, None),
    ("차징 샷", state, None),
    ("아토믹 애로우", state, None),
    ("애로우 샤워", state, None),
    ("애로우 해일", state, None),
    ("아토믹 애로우", state, None),
    ("스나이프", state, None),
    ("샤프 슈터", state, None),
    ("차징 샷", state, None),
]

calculate(
    job_name="hawkeye",
    internal_stat=InternalStat(
        weapon_att=31392, stat_main=133748, crit=607, special=56, swift=1756
    ),
    weapon_pdamage=23.45,
    setitem_state=[
        ("악몽", 2, 0),
        ("지배", 4, 0),
    ],
    engraving_state=[
        ("원한", 3),
        ("예리한 둔기", 3),
        ("타격의 대가", 3),
        ("아드레날린", 3),
        ("정밀 단도", 3),
    ],
    card_state=["남겨진 바람의 절벽 (12)"],
    skill_state={
        "스나이프": SkillState(
            level=12,
            gem=7,
            tripod={
                "약점 포착": 5,
                "손쉬운 먹잇감": 5,
            },
        ),
        "샤프 슈터": SkillState(
            level=12,
            gem=7,
            tripod={
                "급소 타격": 5,
                "약점 포착": 5,
                "집중 사격": 5,
            },
        ),
        "차징 샷": SkillState(
            level=12,
            gem=5,
            tripod={
                "더블 샷": 4,
                "즉발": 4,
            },
        ),
        "애로우 해일": SkillState(
            level=12,
            gem=7,
            tripod={
                "강화된 화살": 4,
                "저속 탄환": 5,
                "웨이브 해일": 5,
            },
        ),
        "아토믹 애로우": SkillState(
            level=12,
            gem=5,
            tripod={
                "피해 증폭": 1,
                "전격 화살": 4,
            },
        ),
        "애로우 샤워": SkillState(
            level=10,
            gem=5,
            tripod={
                "지속력 강화": 4,
                "장대비": 4,
            },
        ),
    },
    enemy=Enemy(armor=6000, reduction=23),
    deal_cycle=deal_cycle,
    cycle_time=62,
)
