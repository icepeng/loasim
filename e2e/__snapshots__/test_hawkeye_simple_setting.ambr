# name: test_hawkeye_simple_setting
  3069532.2804473094
---
